### ISSUE TITLE
Const generics with TypeName derive macro cause compile errors

### ISSUE THREAD

**Issue by edfraenkel** [rkyv/rkyv#383]
This compile time error has the same root cause as issue #374. It can be triggered by the following code:
```rust
#[derive(TypeName)]
struct Struct<const N: usize> {
    a: u32,
    b: u128,
}
```
which results in:
```
error: proc-macro derive panicked
   --> rkyv_typename\tests\test_derive.rs:170:14
    |
170 |     #[derive(TypeName)]
    |              ^^^^^^^^
    |
    = help: message: called `Option::unwrap()` on a `None` value
```

**Comment by edfraenkel** [comment]
Unlike issue #374 it seems like this issue require writing some additional functionality and would influence some design decisions. There are several possible solutions:

1) _Make the const values part of the default type name:_

Implement it such that a unique representation of the constant occurs in the typename. For the currently supported const-generic types this could be implemented with the `Display` trait. However, despite the fact that this is easy to do, I am not sure that this is the right course of afction because of the following considerations.
In a sense, it is 'accidental' that the `Display` trait works because in general, generating unique const identifiers, is not what it is intended for.
In other words: despite the fact that, at this moment, the Display trait would work nicely, it may break for future versions of rust. Future supported const-generic-types may not use the Display trait at all or may not use the Display trait such that a unique output is generated for every value.

2) _Make a minimal implementation for const generics as part of the type name and extend it as the language is extended:_

Forward compatibility could be mitigated by performing compile time type-checks (probably by using some custom private trait). I am pretty sure this is possible, also with human readable error messages.

3) _Use the `Hash` trait to generate the the automatic derive when const generics are involved_

As far as cosistency is concerned this would seem an option that covers more possible const-generic values. It is likely that many (if not all) new const-generic-types will (allow) the implementation of this trait and it could generate a virtually guaranteed unique typename for each const value. However this would be at the expense of readability of the typename and would imply an extra trait bound when const-generics are used. 

4) _Omit const generics from the type name just like lifetime generics can be omitted_

This would not be a viable solution because const generics, unlike generic lifetimes, can affect the serialization/deserialization process. Also, some const generics are already implemented for built-in types, like fixed length arrays. 

5) _Forbid using the automatic derive for const generics_

In this case all implementations for TypeName would need to be hand-written. Allowing them only in such a limited case might not be derilable because it would require the user to write more boiler plate.

**Comment by djkoloski** [comment]
I think the best approach for solving this problem would be to require const generic parameters to implement some new trait like:

```rust
pub unsafe trait ConstGeneric {
    fn build_value<F: FnMut(&str)>(f: F);
}
```

and invoke that directly on the const generic value. The trait can have a safety requirement that distinct values must pass distinct strings to `f`. We can provide implementations for builtin types for now and expand the implementations as required.

**Comment by edfraenkel** [comment]
Seems like a good solution, except perhaps for the unsafe part. Maybe I am overlooking something but I am not sure why the trait would have to be unsafe: implementations of the TypeName trait can already produce typenames that are not unique. In fact the documentation states that there is no uniqueness guarantee. So unless the TypeName trait is made unsafe as well I do not see a reason to make the proposed ConstGeneric trait unsafe.

**Comment by djkoloski** [comment]
Ah you're right, the trait would not need to be `unsafe`.

**Comment by edfraenkel** [comment]
Ok, so I implemented it in my own branch but I realized that the future std traits and the type system of the compiler may require changes to the public interface of the ConstGeneric trait. I didn't delve too deeply into it, but see for instance the discussion about the currently unstable `Structural*Eq` traits which could play a role in future const generics. 

https://hackmd.io/@rust-const-generics/BJ5qbBLod

Depending, for instance, on how the more general const generics are to be implemented it may be necessary to rely on some of these structural trait bounds. 

I would have sugested to make the ConstGeneric trait private as long as only limited const generics are supported, but making it private is tricky because the proc-macro is in another crate.  So in stead, I documented that it may change in the future and I made it hidden in the docs. In any case, since the project is still in the 0.x.y phase some breakage of the public API is be permissible I guess.
